File: src/sc_neurocore/neurons/models/wilson_cowan.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Wilson-Cowan 1972 — excitatory/inhibitory population

from __future__ import annotations

import math
import operator
from dataclasses import dataclass
from typing import SupportsIndex, cast

import numpy as np
import numpy.typing as npt

_STATE_NAMES = ("e", "i")
_PARAM_NAMES = ("w_ee", "w_ei", "w_ie", "w_ii", "tau_e", "tau_i", "a", "theta", "dt")
_NON_NEGATIVE_PARAMS = ("w_ee", "w_ei", "w_ie", "w_ii")
_STRICTLY_POSITIVE_PARAMS = ("tau_e", "tau_i", "a", "dt")
_MAX_STEPS = (1 << 31) - 1
_BACKENDS = ("auto", "python", "rust", "julia", "go", "mojo")


def _step_count(value: object) -> int:
    """Return a C-ABI-safe non-negative batch length."""
    if isinstance(value, bool):
        raise ValueError("n_steps must be a non-negative integer")
    try:
        converted = operator.index(cast(SupportsIndex, value))
    except TypeError as exc:
        raise ValueError("n_steps must be a non-negative integer") from exc
    result = int(converted)
    if not 0 <= result <= _MAX_STEPS:
        raise ValueError(f"n_steps must be in [0, {_MAX_STEPS}]")
    return result


@dataclass
class WilsonCowanUnit:
    """Normalised Wilson-Cowan excitatory/inhibitory population-rate model.

    τ_e dE/dt = -E + S(w_ee·E - w_ei·I + I_ext)
    τ_i dI/dt = -I + S(w_ie·E - w_ii·I)
    S(x) = 1/(1 + exp(-a(x-θ))) - 1/(1 + exp(aθ))

    This maintained reduction omits the original availability/refractory
    factors and an independent inhibitory external drive. It preserves the
    coupled E/I population structure and shifted sigmoid, and advances both
    continuous rates with fixed-step RK4.
    """

    e: float = 0.1
    i: float = 0.05
    w_ee: float = 10.0
    w_ei: float = 6.0
    w_ie: float = 10.0
    w_ii: float = 1.0
    tau_e: float = 1.0
    tau_i: float = 2.0
    a: float = 1.2
    theta: float = 4.0
    dt: float = 0.1

    def __post_init__(self) -> None:
        self._validate_configuration(coerce=True)

    def _validate_configuration(self, *, coerce: bool = False) -> None:
        for name in (*_STATE_NAMES, *_PARAM_NAMES):
            value = float(getattr(self, name))
            if not math.isfinite(value):
                raise ValueError(f"{name} must be finite")
            if coerce:
                setattr(self, name, value)
        for name in _NON_NEGATIVE_PARAMS:
            if getattr(self, name) < 0.0:
                raise ValueError(f"{name} must be non-negative")
        for name in _STRICTLY_POSITIVE_PARAMS:
            if getattr(self, name) <= 0.0:
                raise ValueError(f"{name} must be positive")
        self._validate_state(self.e, self.i)

    @staticmethod
    def _logistic(z: float) -> float:
        if z >= 0.0:
            return 1.0 / (1.0 + math.exp(-z))
        exp_z = math.exp(z)
        return exp_z / (1.0 + exp_z)

    def _validate_rate(self, name: str, value: float) -> float:
        rate = float(value)
        baseline = self._logistic(-self.a * self.theta)
        lower = -baseline
        upper = 1.0
        if not math.isfinite(rate) or rate < lower or rate > upper:
            raise FloatingPointError(f"{name} rate must remain in the Wilson-Cowan state envelope")
        return rate

    def _validate_state(self, e: float, i: float) -> tuple[float, float]:
        return self._validate_rate("e", e), self._validate_rate("i", i)

    def _sigmoid(self, x: float) -> float:
        # Published Wilson-Cowan 1972 two-term sigmoid:
        #   S(x) = 1/(1+exp(-a(x-θ))) − 1/(1+exp(aθ))
        # The subtracted baseline makes S(0) = 0 exactly. The earlier
        # one-term implementation left an artefactual bias
        # `1/(1+exp(aθ))` ≈ 0.008 at x = 0 that shifted the model's
        # fixed points and suppressed the Hopf oscillator regime.
        # math.exp on scalars keeps the per-step cost ~0.7 µs.
        drive = float(x)
        if not math.isfinite(drive):
            raise ValueError("sigmoid input must be finite")
        return self._logistic(self.a * (drive - self.theta)) - self._logistic(-self.a * self.theta)

    def _derivatives(self, e: float, i: float, drive: float) -> tuple[float, float]:
        se = self._sigmoid(self.w_ee * e - self.w_ei * i + drive)
        si = self._sigmoid(self.w_ie * e - self.w_ii * i)
        de = (-e + se) / self.tau_e
        di = (-i + si) / self.tau_i
        if not math.isfinite(de) or not math.isfinite(di):
            raise FloatingPointError("Wilson-Cowan derivative must remain finite")
        return de, di

    def step(self, ext_input: float = 0.0) -> float:
        """Advance one finite excitatory-drive sample and return the E rate."""
        drive = float(ext_input)
        if not math.isfinite(drive):
            raise ValueError("external input must be finite")

        self._validate_configuration()
        e, i = self._validate_state(self.e, self.i)

        k1_e, k1_i = self._derivatives(e, i, drive)
        k2_e, k2_i = self._derivatives(
            e + 0.5 * self.dt * k1_e,
            i + 0.5 * self.dt * k1_i,
            drive,
        )
        k3_e, k3_i = self._derivatives(
            e + 0.5 * self.dt * k2_e,
            i + 0.5 * self.dt * k2_i,
            drive,
        )
        k4_e, k4_i = self._derivatives(
            e + self.dt * k3_e,
            i + self.dt * k3_i,
            drive,
        )
        next_e = e + self.dt * (k1_e + 2.0 * k2_e + 2.0 * k3_e + k4_e) / 6.0
        next_i = i + self.dt * (k1_i + 2.0 * k2_i + 2.0 * k3_i + k4_i) / 6.0
        self.e, self.i = self._validate_state(next_e, next_i)
        return self.e

    def simulate(
        self,
        n_steps: int,
        current: float = 1.5,
        backend: str = "auto",
    ) -> tuple[npt.NDArray[np.float64], npt.NDArray[np.float64]]:
        """Return atomic post-step E/I traces through one maintained backend.

        Explicit unavailable backends fail closed. This instance commits its
        two final rates only after the complete result passes shape, finiteness,
        range, and trace/final consistency checks.
        """
        steps = _step_count(n_steps)
        drive = float(current)
        if not math.isfinite(drive):
            raise ValueError("current must be finite")
        self._validate_configuration()
        if backend not in _BACKENDS:
            raise ValueError(f"backend must be auto/python/rust/julia/go/mojo, got {backend!r}")

        from sc_neurocore.accel import wilson_cowan as backends

        selected = backends.auto_backend() if backend == "auto" else backend
        if selected == "python":
            result = self._simulate_python(steps, drive)
        else:
            if not backends.backend_available(selected):
                raise RuntimeError(f"{selected.title()} WilsonCowan backend is unavailable.")
            runner = {
                "rust": backends.simulate_rust,
                "julia": backends.simulate_julia,
                "go": backends.simulate_go,
                "mojo": backends.simulate_mojo,
            }[selected]
            result = runner(
                self.e,
                self.i,
                self.w_ee,
                self.w_ei,
                self.w_ie,
                self.w_ii,
                self.tau_e,
                self.tau_i,
                self.a,
                self.theta,
                self.dt,
                steps,
                drive,
            )
        e_trace, i_trace, final_e, final_i = backends.normalise_result(
            *result,
            n_steps=steps,
            initial_e=self.e,
            initial_i=self.i,
            a=self.a,
            theta=self.theta,
        )
        self.e, self.i = final_e, final_i
        return e_trace, i_trace

    def _simulate_python(
        self,
        n_steps: int,
        current: float,
    ) -> tuple[npt.NDArray[np.float64], npt.NDArray[np.float64], float, float]:
        """Evaluate a complete local batch without mutating this instance."""
        local = WilsonCowanUnit(
            e=self.e,
            i=self.i,
            w_ee=self.w_ee,
            w_ei=self.w_ei,
            w_ie=self.w_ie,
            w_ii=self.w_ii,
            tau_e=self.tau_e,
            tau_i=self.tau_i,
            a=self.a,
            theta=self.theta,
            dt=self.dt,
        )
        e_trace = np.empty(n_steps, dtype=np.float64)
        i_trace = np.empty(n_steps, dtype=np.float64)
        for index in range(n_steps):
            local.step(current)
            e_trace[index] = local.e
            i_trace[index] = local.i
        return e_trace, i_trace, local.e, local.i

    def reset(self) -> None:
        """Reset the two rates while preserving the configured dynamics."""
        self.e, self.i = 0.1, 0.05

File: src/sc_neurocore/accel/go/services/wilson_cowan.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SC-NeuroCore — Go service for wilson_cowan

package services

import (
	"errors"
	"math"
)

// WilsonCowanUnitState holds the neuron state
type WilsonCowanUnitState struct {
	E     float64
	I     float64
	WEe   float64
	WEi   float64
	WIe   float64
	WIi   float64
	TauE  float64
	TauI  float64
	A     float64
	Theta float64
	Dt    float64
}

// NewWilsonCowanUnit creates a new WilsonCowanUnit neuron with default parameters
func NewWilsonCowanUnit() *WilsonCowanUnitState {
	return &WilsonCowanUnitState{
		E:     0.1,
		I:     0.05,
		WEe:   10.0,
		WEi:   6.0,
		WIe:   10.0,
		WIi:   1.0,
		TauE:  1.0,
		TauI:  2.0,
		A:     1.2,
		Theta: 4.0,
		Dt:    0.1,
	}
}

func finiteWilsonCowan(value float64) bool {
	return !math.IsNaN(value) && !math.IsInf(value, 0)
}

func logisticWilsonCowan(z float64) float64 {
	if z >= 0.0 {
		return 1.0 / (1.0 + math.Exp(-z))
	}
	expZ := math.Exp(z)
	return expZ / (1.0 + expZ)
}

func finiteWilsonCowanRate(value float64, a float64, theta float64) bool {
	baseline := logisticWilsonCowan(-a * theta)
	return finiteWilsonCowan(value) && value >= -baseline && value <= 1.0-baseline
}

func sigmoidWilsonCowan(a float64, theta float64, x float64) (float64, error) {
	if !finiteWilsonCowan(a) || !finiteWilsonCowan(theta) || !finiteWilsonCowan(x) {
		return 0, errors.New("invalid Wilson-Cowan sigmoid input")
	}
	return logisticWilsonCowan(a*(x-theta)) - logisticWilsonCowan(-a*theta), nil
}

func wilsonCowanDerivatives(s *WilsonCowanUnitState, e float64, i float64, iExt float64) (float64, float64, error) {
	se, err := sigmoidWilsonCowan(s.A, s.Theta, s.WEe*e-s.WEi*i+iExt)
	if err != nil {
		return 0, 0, err
	}
	si, err := sigmoidWilsonCowan(s.A, s.Theta, s.WIe*e-s.WIi*i)
	if err != nil {
		return 0, 0, err
	}
	de := (-e + se) / s.TauE
	di := (-i + si) / s.TauI
	if !finiteWilsonCowan(de) || !finiteWilsonCowan(di) {
		return 0, 0, errors.New("invalid Wilson-Cowan derivative")
	}
	return de, di, nil
}

// ValidateWilsonCowan checks the runtime rate state and numerical parameters.
func ValidateWilsonCowan(s *WilsonCowanUnitState) bool {
	if s == nil {
		return false
	}
	return finiteWilsonCowan(s.WEe) && s.WEe >= 0.0 &&
		finiteWilsonCowan(s.WEi) && s.WEi >= 0.0 &&
		finiteWilsonCowan(s.WIe) && s.WIe >= 0.0 &&
		finiteWilsonCowan(s.WIi) && s.WIi >= 0.0 &&
		finiteWilsonCowan(s.TauE) && s.TauE > 0.0 &&
		finiteWilsonCowan(s.TauI) && s.TauI > 0.0 &&
		finiteWilsonCowan(s.A) && s.A > 0.0 &&
		finiteWilsonCowan(s.Theta) &&
		finiteWilsonCowan(s.Dt) && s.Dt > 0.0 &&
		finiteWilsonCowanRate(s.E, s.A, s.Theta) &&
		finiteWilsonCowanRate(s.I, s.A, s.Theta)
}

// Step advances the neuron by one timestep
func (s *WilsonCowanUnitState) Step(iExt float64) (float64, error) {
	if !ValidateWilsonCowan(s) {
		return 0, errors.New("invalid Wilson-Cowan runtime state")
	}
	if !finiteWilsonCowan(iExt) {
		return 0, errors.New("invalid Wilson-Cowan external input")
	}

	k1E, k1I, err := wilsonCowanDerivatives(s, s.E, s.I, iExt)
	if err != nil {
		return 0, err
	}
	k2E, k2I, err := wilsonCowanDerivatives(s, s.E+0.5*s.Dt*k1E, s.I+0.5*s.Dt*k1I, iExt)
	if err != nil {
		return 0, err
	}
	k3E, k3I, err := wilsonCowanDerivatives(s, s.E+0.5*s.Dt*k2E, s.I+0.5*s.Dt*k2I, iExt)
	if err != nil {
		return 0, err
	}
	k4E, k4I, err := wilsonCowanDerivatives(s, s.E+s.Dt*k3E, s.I+s.Dt*k3I, iExt)
	if err != nil {
		return 0, err
	}
	nextE := s.E + s.Dt*(k1E+2.0*k2E+2.0*k3E+k4E)/6.0
	nextI := s.I + s.Dt*(k1I+2.0*k2I+2.0*k3I+k4I)/6.0
	if !finiteWilsonCowanRate(nextE, s.A, s.Theta) || !finiteWilsonCowanRate(nextI, s.A, s.Theta) {
		return 0, errors.New("invalid Wilson-Cowan candidate state")
	}

	s.E = nextE
	s.I = nextI
	return s.E, nil
}

// SimulateWilsonCowanUnit runs the neuron for n steps
func SimulateWilsonCowanUnit(nSteps int, iExt float64) ([]float64, int) {
	s := NewWilsonCowanUnit()
	trace := make([]float64, nSteps)
	spikes := 0
	for t := 0; t < nSteps; t++ {
		_, err := s.Step(iExt)
		if err != nil {
			trace[t] = math.NaN()
			continue
		}
		trace[t] = s.E
	}
	return trace, spikes
}
